modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java

```diff
@@ -285,6 +293,79 @@
         this.logTxRecords = IgniteSystemProperties.getBoolean(IGNITE_WAL_LOG_TX_RECORDS, false);
     }
 
+    /** {@inheritDoc} */
+    @Override public void onActivate(GridKernalContext kctx) throws IgniteCheckedException {
+        IgniteCacheDatabaseSharedManager psMgr = kctx.cache().context().database();
+
+        if (psMgr instanceof GridCacheDatabaseSharedManager) {
+            GridCacheDatabaseSharedManager psMgr0 = (GridCacheDatabaseSharedManager)psMgr;
+
+            psMgr0.addCheckpointListener(this);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onDeActivate(GridKernalContext kctx) {
+        IgniteCacheDatabaseSharedManager psMgr = kctx.cache().context().database();
+
+        if (psMgr instanceof GridCacheDatabaseSharedManager) {
+            GridCacheDatabaseSharedManager psMgr0 = (GridCacheDatabaseSharedManager)psMgr;
+
+            psMgr0.removeCheckpointListener(this);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onCheckpointBegin(Context ctx) throws IgniteCheckedException {
+        CheckpointRecord rec = ctx.checkpointRecord();
+
+        FileWALPointer pntCandidate = null;
+
+        FileWALPointer systemPtrCandidate = findCandidate(sysThreadMap.values());
+        FileWALPointer userPtrCandidate = findCandidate(idMap.values());
+
+        if (systemPtrCandidate == null && userPtrCandidate == null)
+            return;
+
+        if (systemPtrCandidate != null && userPtrCandidate == null)
+            rec.checkpointMark(systemPtrCandidate);
+        else if (systemPtrCandidate == null && userPtrCandidate != null)
+            rec.checkpointMark(userPtrCandidate);
+        else if (userPtrCandidate.compareTo(systemPtrCandidate) < 0)
+            rec.checkpointMark(userPtrCandidate);
+        else
+            rec.checkpointMark(systemPtrCandidate);
+    }
+
+    /**
+     *
+     */
+    private FileWALPointer findCandidate(Collection<IgniteInternalTx> txs) {
+        FileWALPointer pntCandidate = null;
+
+        for (IgniteInternalTx tx : txs)
+            if (tx instanceof IgniteTxAdapter) {
+                IgniteTxAdapter txAdapter = (IgniteTxAdapter)tx;
+
+                WALPointer ptr = txAdapter.prepPtr;
+
+                if (ptr != null && ptr instanceof FileWALPointer) {
+                    FileWALPointer ptr0 = (FileWALPointer)ptr;
```

Comment:
Also, do not cast to FileWALPointer here, add Comparable to wal pointer interface.